### ISSUE TITLE
Unsafe DllMain usage

### ISSUE THREAD

**Issue by jermarti** [oneapi-src/oneTBB#173]
I left a summary of this on the [tbb forums](https://software.intel.com/en-us/forums/intel-threading-building-blocks/topic/815753) last week, but it hasn't got any traction so i thought i'd cross post here too.

The issue is the current use of callbacks from the scheduler framework (at least when using flow graphs) is unsafe for many operations, and even runs against safe DllMain practices listed on [this](https://software.intel.com/en-us/articles/the-ultimate-question-of-programming-refactoring-and-everything) Intel blog post.

Here's an example stack that eventually calls into a continue node callback:

tbb_debug.dll!tbb::internal::custom_scheduler<tbb::internal::IntelSchedulerTraits>::local_wait_for_all(tbb::task & parent, tbb::task * child) Line 517    C++
tbb_debug.dll!tbb::internal::generic_scheduler::cleanup_master(bool blocking_terminate) Line 1317    C++
tbb_debug.dll!tbb::internal::governor::auto_terminate(void * arg) Line 222    C++
tbb_debug.dll!tbb::internal::governor::terminate_auto_initialized_scheduler() Line 139    C++
tbb_debug.dll!DllMain(void * __formal, unsigned long reason, void * lpvReserved) Line 264    C++

In short, it appears that an invocation of DllMain can call back into user callbacks on graph nodes, which is unsafe because while a DllMain is open a loader lock is established in windows. There's an entire family of operations on windows that can deadlock if done while inside a loader lock, and that blog post has a pretty good list (republished here):

- Call LoadLibrary or LoadLibraryEx (either directly or indirectly). This can cause a deadlock or a crash.

- Call GetStringTypeA, GetStringTypeEx, or GetStringTypeW (either directly or indirectly). This can cause a deadlock or a crash.

- Synchronize with other threads. This can cause a deadlock.

- Acquire a synchronization object that is owned by code that is waiting to acquire the loader lock. This can cause a deadlock.

- Initialize COM threads by using CoInitializeEx Under certain conditions, this function can call LoadLibraryEx.

- Call the registry functions. These functions are implemented in Advapi32.dll. If Advapi32.dll is not initialized before your DLL, the DLL can access uninitialized memory and cause the process to crash.

- Call CreateProcess. Creating a process can load another DLL.

- Call ExitThread. Exiting a thread during DLL detach can cause the loader lock to be acquired again, causing a deadlock or a crash.

- Call CreateThread. Creating a thread can work if you do not synchronize with other threads, but it is risky.

- Create a named pipe or other named object (Windows 2000 only). In Windows 2000, named objects are provided by the Terminal Services DLL. If this DLL is not initialized, calls to the DLL can cause the process to crash.

- Use the memory management function from the dynamic C Run-Time (CRT). If the CRT DLL is not initialized, calls to these functions can cause the process to crash.

- Call functions in User32.dll or Gdi32.dll. Some functions load another DLL, which may not be initialized.

- Use managed code.

I'm currently running into this in a project that needs to call back into managed code on a per node basis. I've confirmed this behavior persists in the TBB 2019 Update 8 release.

Optimally the solution here would be for DllMain to become a stub that tracks lifetime, and doesn't stay open due to the unsafe rules around window's loader lock. I realize that may not be possible and other solutions could be done too, and i'm very open to work arounds for this issue.

At the very least if there's some reason why TBB needs to both keep DllMain open AND call back into user code from it, that should be documented as an important consideration for developers evaluating the scheduler framework.

**Comment by alexey-katranov** [comment]
Normally, the clean up code in DLLMain should not call user callbacks. However, it might be caused with "forgotten" TBB tasks if wait_for_all is not called for some reason. When using flow graph, do you call `graph::wait_for_all`?

**Comment by jermarti** [comment]
Hi!
I do call wait_for_all. The stack did look a little odd to me as this fires almost immediately when starting processing, and as you pointed out it appears to be in last ditch cleanup code. 
My usage for the graph is as follows:

- Create a graph with a tbb::task_group_context (using the default ctor)
- Create a start node (which is a tbb::flow::broadcast_node<tbb::flow::continue_msg>)
- Populate the graph with these nodes: 
tbb::flow::multifunction_node<tbb::flow::continue_msg, tbb::flow::tuple<tbb::flow::continue_msg, tbb::flow::continue_msg>, tbb::flow::lightweight>

tbb::flow::continue_node<tbb::flow::continue_msg> 

tbb::flow::async_node<tbb::flow::continue_msg, tbb::flow::continue_msg>

- Start the graph by running this:
m_startNode->try_put(tbb::flow::continue_msg());
m_graph->wait_for_all();

And almost immediately after i get a loader lock exception when the flow scheduler calls back into a continue node as described above. This happens after wait_for_all() is called.

I've tried doing an explicit LoadLibrary on tbb.dll / tbb_debug.dll (depending on compile mode) before interacting with the tbb::flow code to increment the refcount on the dll, and that has no effect. 

This probably wont matter, but just for context my project is a fully native (no clr) x64 dll compiled in MSVC 2019, linked to the vc14 libs / dlls in the precompiled download package for tbb.

**Comment by alexey-katranov** [comment]
Do you use [`async_node_type::gateway_type::reserve_wait`](https://www.threadingbuildingblocks.org/docs/help/index.htm#reference/flow_graph/async_node_cls.html) to notify the graph about the asynchronous activity?

**Comment by jermarti** [comment]
Yup, here's the callback function that runs in every case, with ProcessAsync() a callback into the actual async code and CancelProcessing() a call to task_group_context::cancel_group_execution() (which isnt firing when i see the loader lock issues):

	[=](const tbb::flow::continue_msg&, TBBAsyncNode::gateway_type& asyncGateway)
	{
		asyncGateway.reserve_wait();
		std::thread([&]
		{

			bool success = false;
			try
			{
				ProcessAsync();
				success = true;
			}
			catch (std::exception ex)
			{
				GetKentGraph().CancelProcessing();
			}

			if (success)
			{
				asyncGateway.try_put(tbb::flow::continue_msg());
				asyncGateway.release_wait();
			}
		}).detach();
	}

**Comment by alexey-katranov** [comment]
You approach seems correct. It should not lead to the reported issue. Is it possible to share the reproducer? Or provide more details about how TBB is used, e.g. callstacks, information about other threads (beside the async threads), the number of graphs, more details about the algorithm structure , if at least once an exception was thrown (even if caught) or any other information that you think might be relevant.

**Comment by jermarti** [comment]
Happy to share as many details as possible, though sharing the code itself is probably impossible because it's tented internal to my company. I may be able to put together a stand alone test project if we run out of options though.

The answer the immediate questions, there's only one graph in the process, and i confirmed there are no exceptions thrown before the loader lock fires by setting a stop instruction in the debugger on any/all c++ exceptions. 
The vast majority of the threads in the process when the loader lock fires are tbb threads from the graph executing, but there are a handful of others which are either the main thread or clr related threads (there is managed code in the process, though the code in question here is all native (which eventually calls into clr code and triggers the loader lock if it's from a DllMain).

I have cloned the latest from the tbb github so i should be able to look at state inside the tbb code. Are there any places you'd suggest i look to investigate what might be causing this? The repro is pretty easy, and happens in debug luckily, so i should be able to also set data breakpoints if there are any variables that would obviously govern this behavior.

I'm a little reluctant to post stack traces here, would it be possible to msg you directly with extended debugging data? I'm looking at the stacks now it it is clear the crt is trying to join the thread the dllmain is associated with while the main thread is calling wait_for_all() on the graph.

**Comment by jermarti** [comment]
This is a full stack of a loader lock firing, which is safe to share for sure. DLL main is being called with reason DLL_THREAD_DETACH:

 	KentStudioPipeline2.dll!StudioKent::KentContinueNode::<lambda>(const tbb::flow::interface10::continue_msg & __formal) Line 233	C++
 	KentStudioPipeline2.dll!tbb::flow::interface10::internal::function_body_leaf<tbb::flow::interface10::continue_msg,tbb::flow::interface10::continue_msg,void <lambda>(const tbb::flow::interface10::continue_msg &) >::operator()(const tbb::flow::interface10::continue_msg & i) Line 147	C++
 	KentStudioPipeline2.dll!tbb::flow::interface10::internal::continue_input<tbb::flow::interface10::continue_msg,tbb::flow::interface10::internal::Policy<void> >::apply_body_bypass(tbb::flow::interface10::continue_msg __formal) Line 747	C++
 	KentStudioPipeline2.dll!tbb::flow::interface10::internal::apply_body_task_bypass<tbb::flow::interface10::internal::continue_input<tbb::flow::interface10::continue_msg,tbb::flow::interface10::internal::Policy<void> >,tbb::flow::interface10::continue_msg>::execute() Line 312	C++
 	tbb_debug.dll!tbb::internal::custom_scheduler<tbb::internal::IntelSchedulerTraits>::local_wait_for_all(tbb::task & parent, tbb::task * child) Line 517	C++
 	tbb_debug.dll!tbb::internal::generic_scheduler::cleanup_master(bool blocking_terminate) Line 1317	C++
 	tbb_debug.dll!tbb::internal::governor::auto_terminate(void * arg) Line 221	C++
 	tbb_debug.dll!tbb::internal::governor::terminate_auto_initialized_scheduler() Line 139	C++
 	tbb_debug.dll!tbb::internal::DllMain(void *, unsigned long reason, void * lpvReserved) Line 264	C++
 	tbb_debug.dll!dllmain_dispatch(HINSTANCE__ * const instance, const unsigned long reason, void * const reserved) Line 195	C++
 	tbb_debug.dll!_DllMainCRTStartup(HINSTANCE__ * const instance, const unsigned long reason, void * const reserved) Line 249	C++
 	ntdll.dll!LdrpCallInitRoutine()	Unknown
 	ntdll.dll!LdrShutdownThread()	Unknown
 	ntdll.dll!RtlExitUserThread()	Unknown
 	KernelBase.dll!FreeLibraryAndExitThread()	Unknown
 	ucrtbased.dll!common_end_thread(const unsigned int return_code) Line 264	C++
 	ucrtbased.dll!_endthreadex(unsigned int return_code) Line 278	C++
 	ucrtbased.dll!thread_start<unsigned int (__cdecl*)(void *),1>(void * const parameter) Line 104	C++
 	kernel32.dll!BaseThreadInitThunk()	Unknown
 	ntdll.dll!RtlUserThreadStart()	Unknown


Meanwhile the main thread is waiting patiently:

>	KentStudioPipeline2.dll!tbb::task::wait_for_all() Line 791	C++
 	KentStudioPipeline2.dll!tbb::flow::interface10::graph::wait_functor::operator()() Line 248	C++
 	KentStudioPipeline2.dll!tbb::interface7::internal::delegated_function<tbb::flow::interface10::graph::wait_functor const ,void>::operator()() Line 91	C++
 	tbb_debug.dll!tbb::interface7::internal::task_arena_base::internal_execute(tbb::interface7::internal::delegate_base & d) Line 943	C++
 	KentStudioPipeline2.dll!tbb::interface7::task_arena::execute_impl<void,tbb::flow::interface10::graph::wait_functor const >(const tbb::flow::interface10::graph::wait_functor & f) Line 214	C++
 	KentStudioPipeline2.dll!tbb::interface7::task_arena::execute<tbb::flow::interface10::graph::wait_functor>(const tbb::flow::interface10::graph::wait_functor & f) Line 354	C++
 	KentStudioPipeline2.dll!tbb::flow::interface10::graph::wait_for_all() Line 342	C++
 	KentStudioPipeline2.dll!StudioKent::KentGraph::WaitForGraphComplete() Line 169	C++

**Comment by jermarti** [comment]
From the source code of ucrt/startup/thread.cpp in the windows SDK 10.0.18362.0, this is the code that triggers the actual thread work:

__try
    {
        ThreadProcedure const procedure = reinterpret_cast<ThreadProcedure>(context->_procedure);
        if constexpr (Ex)
        {
            _endthreadex(procedure(context->_context));
        }
        else
        {
            procedure(context->_context);
            _endthreadex(0);
        }
    }
    __except (_seh_filter_exe(GetExceptionCode(), GetExceptionInformation()))
    {
        // Execution should never reach here:
        _exit(GetExceptionCode());
    }

Given we're in _endthreadex() it looks like the only way for that to trigger is if the actual thread procedure has been called and returned

**Comment by alexey-katranov** [comment]
Thank you for the information. Currently, I do not have any additional thoughts but I will continue to investigate the case.
As for confidential information, you can use ‘inteltbbdevelopers(at)intel.com’ not to share it on github.

**Comment by jermarti** [comment]
Hi!
Just an update on my side. I confirmed if i make a modified version of TBB where the last ditch cleanup is never run on thread detach then the loaderlock is never triggered. This seems to be the same cleanup close fired on process detach that triggers emergency cleanup.

It seems that the underlying assumption of tbb is either that threads should never join while the scheduler is active, or the heavy handed cleanup on a per thread basis needs to be refcounted so it only fires on the last thread joining. The stacks i'm observing are triggering cleanup code on every thread join, which is just grabbing more work before existing but doing so while in the context of the DllMain loader lock.

**Comment by nofuturre** [comment]
@jermarti is this issue still relevant?
